File: 113__Path_Sum_II.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        def helper(root, val, curr_path):
            if root is None:
                return
            path = curr_path + [root.val]
            if root.left is None and root.right is None:
                if root.val == val:
                    result.append(path)
                return
            helper(root.left, val - root.val, path)
            helper(root.right, val - root.val, path)
        result = []
        helper(root, sum, [])
        return result


    

File: Java/113__Path_Sum_II.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {

    List<List<Integer>> result;

    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        result = new ArrayList<>();
        List<Integer> list = new ArrayList<Integer>();
        helper(list, root, sum);
        return result;
    }

    private void helper(List<Integer> list, TreeNode root, int sum) {
        if (root == null) return;
        sum = sum - root.val;
        list.add(root.val);
        if (root.left == null && root.right == null && sum == 0)
            result.add(new ArrayList<Integer>(list));
        if (root.left != null) {
            helper(list, root.left, sum);
            list.remove(list.size() - 1);
        }
        if (root.right != null) {
            helper(list, root.right, sum);
            list.remove(list.size() - 1);
        }
    }
}
